src/rules.jl

```diff
@@ -116,3 +123,55 @@ rrule(::Any, ::Vararg{Any}) = nothing
 # explicitly empty body to save the compiler that work.
 
 (::Core.kwftype(typeof(rrule)))(::Any, ::Any, ::Vararg{Any}) = nothing
+
+"""
+    RuleConfig{F, R, T}
+
+The configuration
+ - `F`: **frule-like**. This is singleton `typeof` a function which acts like `frule`, but
+   which functions via invoking an AD system. it must match the [`frule`](@ref) signature.
+   If you do not have such a function it must be set to `Nothing` instead.
+ - `R`: **rrule-like**. This is singleton `typeof` a function which acts like `rrule`, but
+   which functions via invoking an AD system. it must match the [`rrule`](@ref) signature.
+   If you do not have such a function it must be set to `Nothing` instead.
+ - `T`: **traits**. This should be a `Union` of all special traits needed for rules to be
+   allowed to be defined for your AD. If nothing special this should be set to `Union{}`.
+
+Rule authors can dispatch on this config when defining rules.
+For example:
+```julia
+# only define rrule for `pop!` on AD systems where mutation is supported.
+rrule(::RuleConfig{<:Any,<:Any,>:SupportsMutation}, typeof(pop!), ::Vector) = ...
+
+# this definition of map is for any AD that defines a forwards mode
+rrule(conf::RuleConfig{<:Function}, typeof(map), ::Vector) = ...
+
+# this definition of map is for any AD that only defines a reverse mode.
+# It is not as good as the rrule that can be used if the AD defines a forward-mode as well.
+rrule(conf::RuleConfig{Nothing,<:Function}, typeof(map), ::Vector) = ...
+```
+
+For more details see [rule configurations and calling back into AD](@ref config).
+"""
+abstract type RuleConfig{F<:Union{Function,Nothing}, R<:Union{Function,Nothing}, T} end
```

Comment:
I suspect we'll need to let people actually pass in objects that contain data.

For example, `Zygote`'s contexts typicaly contain information about any globals that the programme being differentiated uses.

Additionally, restricting this to be a `Function` feels a bit restrictive -- I'd be inclined let it be anything.

Reply:
The reason it is a function is two-fold.

### 1 Dispatch
We need to dispatch only when it is not Not `Nothing`, and the simplest way is to insist that it *is* something else.
So we might want ti dispatch to have a rule that is only defines if we can't call into forward mode but can call into reverse.
That would look like `rrule(::RuleConfig{Nothing, <:Function}, ...)`
 
### 2 invocation.
Doing it this way means we can use `F.instance` to get the singleton function `f` from `F=typeof(f)`, and thus can define the `frule_via_ad` and `rrule_via_ad` based on only that type parameter.
Note that this strictly requires a actual function `Function` not any other thing subtyping it.

## Alternatives:
we could require that all implementors provide the overloads to `frule_via_ad` and `rrule_via_ad` as appropriate (thus solving **invocation**).
and find some other way of signally that they support it.

Two options for that would be just the special functionality  `T` parameter (which would be our only parameter.
Bearing in mind that we need to be able to dispatch both on what it has, and what it doesn't have.

Either we require all to provide a one of either `NoForwards`/`HasForwards`, and `NoReverse`/`HasReverse`,
and dispatch via `rrule(::Config{>:Union{NoForwards, HasReverse}, ...)`
We need them to provide both in this case since we can't have  `NoForwards` and `NoReverse` as defaults or we would run into ambiguity errors that can't be solved via `>:`

Or we have a single trait that it itself parameterised: `HasModes{F, R}` where `F` and `R` are bools, and dispatch via `rrule(::Config{>:HasModes{false,true}}, ...)`.
Which has the advantage that now we can treat having neither as the default.


Do you want to have a quick call about this some time today? I am free whenever